spec/controllers/books_controller_spec.rb

```diff
@@ -0,0 +1,181 @@
+      expect(subject.status).to eq(200)
+      expect(assigns(:book)).to be_a_kind_of(Book)
+    end
+  end
+
+  describe 'POST /create' do
+    let(:book_attrs) { attributes_for(:book) }
+
+    subject { post :create, params: { book: book_attrs } }
+
+    context 'valid params' do
+      it 'creates a new post' do
+        expect(subject).to redirect_to(books_path)
+      end
+    end
+
+    context 'invalid params' do
+      before do
+        book_attrs[:name] = nil
+      end
+      it 'doesnt create a new post' do
+        bad_book = build(:book, book_attrs)
+        bad_book.save
+        expect(subject).to redirect_to(new_book_path(errors: bad_book.errors.full_messages))
+      end
+    end
+  end
+
+  describe 'GET /edit' do
+    subject { get :edit, params: }
+
+    let(:book) { create(:book) }
+
+    context 'valid id' do
+      let(:params) { { id: book.id } }
+
+      it 'sets book correctly' do
+        expect(subject.status).to eq(200)
+        expect(assigns(:book)).to eq(Book.find(book.id))
+      end
+    end
+
+    context 'invalid id' do
+      let(:params) { { id: -1 } }
+
+      it 'raises error' do
+        expect { subject }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+  end
+
+  describe 'PATCH /update' do
+    subject { patch :update, params: }
+
+    let(:book) { create(:book) }
+    let(:book_attrs) { attributes_for(:book) }
+
+    context 'valid target id' do
+      let(:params) { { id: book.id, book: book_attrs } }
+
+      context 'valid book attrs' do
+        it 'updates book correctly' do
+          expect(subject).to redirect_to(book_path(book))
+          expect { book.reload }.to change(book, :name).from(book.name)
+                                                       .to(book_attrs[:name])
+                               .and change(book, :description).from(book.description)
+                                                              .to(book_attrs[:description])
+        end
+      end
+
+      context 'invalid book attrs' do
+        before { book_attrs[:name] = nil }
+
+        it 'redirects errors' do
+          book.update(book_attrs)
+          expect(subject).to redirect_to(edit_book_path(book, errors: book.errors.full_messages))
+        end
+      end
+    end
+
+    context 'invalid target id' do
+      let(:params) { { id: -1 } }
+
+      it 'raises error' do
+        expect { subject }.to raise_error(ActiveRecord::RecordNotFound)
+      end
+    end
+  end
+
+  describe 'DELETE /destroy' do
+    subject { delete :destroy, params: }
+
+    let(:book_a) { create(:book) }
+    let(:book_b) { create(:book) }
+    let(:book_c) { create(:book) }
+    let(:review_a) { create(:review, book: book_b) }
+    let(:review_b) { create(:review, book: book_b) }
+    let(:review_c) { create(:review, book: book_c) }
+
+    context 'valid target id' do
+      context 'book with no reviews' do
+        let(:params) { { id: book_a.id } }
+
+        before do
+          book_a
+        end
```

Comment:
optional: We can use this if it only 1 line
```suggestion
        before { book_a }
```